File: chamber-library/src/main/java/com/dtp/TableBuilder.java
package com.dtp;

import com.dtp.columns.BooleanColumn;
import com.dtp.columns.Column;
import com.dtp.columns.DoubleColumn;
import com.dtp.columns.FloatColumn;
import com.dtp.columns.IntColumn;
import com.dtp.columns.LongColumn;
import com.dtp.columns.StringColumn;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by ner on 4/6/17.
 */

public class TableBuilder {

    private static final String TEXT = "TEXT";
    private static final String INTEGER = "INTEGER";
    private static final String REAL = "REAL";
    private static final String BOOLEAN = "BOOLEAN";
    private static final String BLOB = "BLOB";

    private static final String SPACE = " ";
    private static final String PERIOD = ".";
    private static final String COMMA = ",";
    private static final String DEFAULT = "DEFAULT";

    private static final String NOT_NULL = "NOT NULL";
    private static final String UNIQUE = "UNIQUE";
    private static final String REFERENCES = "REFERENCES";

    private StringBuilder createStringBuilder = new StringBuilder();
    private List<String> foreignKeys = new ArrayList<String>();
    private String currentTable;
    private List<String> columnNames = new ArrayList<>();

    public String generateCreateString(String tableName, Column[] columns) {
        open(tableName);

        for (Column column : columns) {
            if (column.name.contains("ChamberId"))
                continue;

            ColumnBuilder columnBuilder = getColumnBuilderForColumn(column);

            if (column.notNull)
                columnBuilder.notNull();

            if (column.unique)
                columnBuilder.unique();

            //TODO default value
            //TODO foreignKey

            columnBuilder.build();
        }

        return generateCreateString();
    }

    private void open(String tableName) {
        prepareForNewTable(tableName);

        createStringBuilder.append("CREATE TABLE ");
        createStringBuilder.append(currentTable);
        createStringBuilder.append(" ( ");
        createStringBuilder.append(tableName);
        createStringBuilder.append("ChamberId");
        createStringBuilder.append(" INTEGER PRIMARY KEY AUTOINCREMENT");

        columnNames.add(currentTable + PERIOD + "CHAMBER_ID");
    }

    private void prepareForNewTable(String tableName) {
        createStringBuilder = new StringBuilder();
        columnNames = new ArrayList<>();
        foreignKeys = new ArrayList<>();
        currentTable = tableName;
    }

    private String generateCreateString() {
        createStringBuilder.append(" ");

        for (String foreignKey : foreignKeys) {
            createStringBuilder.append(COMMA);
            createStringBuilder.append(SPACE);
            createStringBuilder.append(foreignKey);
        }

        createStringBuilder.append(")");

        return createStringBuilder.toString();
    }

    private ColumnBuilder getColumnBuilderForColumn(Column column) {
        if (column instanceof StringColumn) {
            return new ColumnBuilder().appendText(column.name);
        } else if (column instanceof IntColumn || column instanceof LongColumn) {
            return new ColumnBuilder().appendInt(column.name);
        } else if (column instanceof FloatColumn || column instanceof DoubleColumn) {
            return new ColumnBuilder().appendReal(column.name);
        } else if (column instanceof BooleanColumn) {
            return new ColumnBuilder().appendBoolean(column.name);
        } else {
            throw new IllegalArgumentException("ColumnBuilder does not support " + column.getClass().getSimpleName());
        }
    }

    class ColumnBuilder {
        private String columnName;
        private String type;
        private List<String> constraints = new ArrayList<>();
        private Object defaultValue;

        ColumnBuilder appendText(String columnName) {
            return append(TEXT, columnName);
        }

        ColumnBuilder appendInt(String columnName) {
            return append(INTEGER, columnName);
        }

        ColumnBuilder appendReal(String columnName) {
            return append(REAL, columnName);
        }

        ColumnBuilder appendBoolean(String columnName) {
            return append(BOOLEAN, columnName);
        }

        public ColumnBuilder appendBlob(String columnName) {
            return append(BLOB, columnName);
        }

        public ColumnBuilder notNull() {
            constraints.add(NOT_NULL);

            return this;
        }

        public ColumnBuilder unique() {
            constraints.add(UNIQUE);

            return this;
        }

        public ColumnBuilder foreignKey(String foreignTableName, String foreignColumnName) {
            foreignKeys.add("FOREIGN KEY (" + columnName + ")" + REFERENCES + foreignTableName + "(" + foreignColumnName + ")");

            return this;
        }

        private ColumnBuilder append(String type, String columnName) {
            this.type = type;
            this.columnName = columnName;

            return this;
        }

        void build() {
            createStringBuilder.append(COMMA);
            createStringBuilder.append(columnName);
            createStringBuilder.append(SPACE);
            createStringBuilder.append(type);

            for (String constraint : constraints) {
                createStringBuilder.append(SPACE);
                createStringBuilder.append(constraint);
            }

            if (defaultValue != null) {
                createStringBuilder.append(DEFAULT);

                if (defaultValue instanceof String || defaultValue instanceof Integer ||
                        defaultValue instanceof Long || defaultValue instanceof Float || defaultValue instanceof Double) {
                    createStringBuilder.append(defaultValue);
                } else if (defaultValue instanceof Boolean) {
                    createStringBuilder.append((Boolean) defaultValue ? 1 : 0);
                }
            }

            columnNames.add(currentTable + PERIOD + columnName);
        }
    }
}

File: chamber-library/src/main/java/com/dtp/TableBuilder.kt
package com.dtp

import com.dtp.columns.*
import com.dtp.columns.Column.Companion.CHAMBER_ID
import java.util.*

/**
 * Created by ner on 4/10/17.
 */

class TableBuilder {

    // Types
    val TEXT = "TEXT"
    val INTEGER = "INTEGER"
    val REAL = "REAL"
    val BOOLEAN = "INTEGER"
    val BLOB = "BLOB"

    val NOT_NULL = "NOT NULL"
    val REFERENCES = "REFERENCES "
    val UNIQUE = "UNIQUE"

    private val SPACE = " "
    private val PERIOD = ""
    private val COMMA = ","
    private val DEFAULT = " DEFAULT "

    private var createString = StringBuilder()

    private var foreignKeys = ArrayList<String>()

    private var currentTable: String? = null

    private var columns = ArrayList<String>()

    fun buildTable(tableName: String, columns: Array<Column>): String {
        open(tableName)

        for (column in columns) {
            if (column.name == CHAMBER_ID)
                continue

            val columnBuilder = getColumnBuilder(column)

            if (column.notNull)
                columnBuilder.notNull()
            if (column.unique)
                columnBuilder.unique()

//            column.defaultValue?.let { columnBuilder.defaultValue(it) }
//
//            column.foreignKey?.let { columnBuilder.foreignKey(it.tableName, it.column) }

            columnBuilder.build()
        }

        return retrieveCreateTableString()
    }

    private fun getColumnBuilder(column: Column): ColumnBuilder {
        when (column) {
            is StringColumn -> return buildTextColumn(column.name)
            is IntColumn, is LongColumn -> return buildIntColumn(column.name)
            is FloatColumn, is DoubleColumn -> return buildRealColumn(column.name)
            is BooleanColumn -> return buildBooleanColumn(column.name)
            else -> throw UnsupportedOperationException("No ColumnBuilder found for type ${column::class.java.simpleName}")
        }
    }

    private fun open(tableName: String): String {
        prepareNewTable(tableName)

        createString.append("CREATE TABLE ")
        createString.append(currentTable)
        createString.append(" ( ")
        createString.append(CHAMBER_ID)
        createString.append(" INTEGER PRIMARY KEY AUTOINCREMENT")

        columns.add("$currentTable$PERIOD$CHAMBER_ID")

        return tableName
    }

    private fun prepareNewTable(tableName: String) {
        createString = StringBuilder()
        columns = ArrayList<String>()
        foreignKeys = ArrayList<String>()
        currentTable = tableName
    }

    private fun buildTextColumn(columnName: String): ColumnBuilder {
        return ColumnBuilder().appendText(columnName)
    }

    private fun buildIntColumn(columnName: String): ColumnBuilder {
        return ColumnBuilder().appendInt(columnName)
    }

    private fun buildRealColumn(columnName: String): ColumnBuilder {
        return ColumnBuilder().appendReal(columnName)
    }

    private fun buildBooleanColumn(columnName: String): ColumnBuilder {
        return ColumnBuilder().appendBoolean(columnName)
    }

    private fun buildBlobColumn(columnName: String): ColumnBuilder {
        return ColumnBuilder().appendBlob(columnName)
    }

    private fun retrieveCreateTableString(): String {
        createString.append(" ")

        for (foreignKey in foreignKeys) {
            createString.append(COMMA)
            createString.append(SPACE)
            createString.append(foreignKey)
        }

        createString.append(")")

        return createString.toString()
    }

    private inner class ColumnBuilder {
        private lateinit var columnName: String

        private var type: String? = null

        private val constraints: MutableList<String>

        private var defaultValue: Any? = null

        init {
            constraints = ArrayList<String>()
        }

        fun appendText(columnName: String): ColumnBuilder {
            this.columnName = columnName
            type = TEXT

            return this
        }

        fun appendInt(columnName: String): ColumnBuilder {
            this.columnName = columnName

            type = INTEGER

            return this
        }

        fun appendReal(columnName: String): ColumnBuilder {
            this.columnName = columnName

            type = REAL

            return this
        }

        fun appendBoolean(columnName: String): ColumnBuilder {
            this.columnName = columnName

            type = BOOLEAN

            return this
        }

        fun appendBlob(columnName: String): ColumnBuilder {
            this.columnName = columnName

            type = BLOB

            return this
        }

        fun notNull(): ColumnBuilder {
            constraints.add(NOT_NULL)

            return this
        }

        fun foreignKey(tableName: String, column: Column): ColumnBuilder {
            foreignKeys.add("FOREIGN KEY (${this.columnName}) $REFERENCES+$tableName(${column.name})")

            return this
        }

        fun unique(): ColumnBuilder {
            constraints.add(" $UNIQUE")

            return this
        }

        fun defaultValue(value: Any): ColumnBuilder {
            defaultValue = value

            return this
        }

        fun build(): String {
            createString.append(COMMA)
            createString.append(columnName)
            createString.append(SPACE)
            createString.append(type)

            for (constraint in constraints) {
                createString.append(SPACE)
                createString.append(constraint)
            }

            defaultValue?.let {
                createString.append(DEFAULT)

                when (it) {
                    is String, is Int, is Long, is Float, is Double -> createString.append(it)
                    is Boolean -> createString.append(if (it) 1 else 0)
                    else -> throw IllegalArgumentException("Default value must be of type String, Int, Long or Boolean. Default value type is ${it.javaClass}")
                }
            }

            columns.add(currentTable + PERIOD + columnName)

            return columnName
        }
    }
}